### PR TITLE
Fix image scan issues

### PR DESCRIPTION
**Pull request by thardeck**
Documentation of image scan: https://fleet.rancher.io/imagescan

---

Use `known_hosts` provided by ssh secret for go-git  the same way it is used for gitjob. This allows SSH based GIT connections.

The SSH private key and `known_hosts` can be provided as `clientSecretName` like this: 
e. g.: `kubectl create -n fleet-local secret generic <CLIENT_SECRET_NAME> --type kubernetes.io/ssh-auth --from-file=ssh-privatekey="<PRIVATE_KEY_PATH>" --from-file=known_hosts="<KNOWN_HOSTS_PATH>"`

Fixes rancher/fleet#797

---
Allow writing to `/tmp` besides having a read only root file system.

Fixes rancher/fleet#1161

### CODE REVIEW

File: charts/fleet/templates/deployment.yaml [hunk 1]
```diff
@@ -56,13 +56,17 @@ spec:
           allowPrivilegeEscalation: false
           readOnlyRootFilesystem: true
         {{- end }}
-        {{- if .Values.cpuPprof }}
         volumeMounts:
+          - mountPath: /tmp
+            name: tmp
+        {{- if .Values.cpuPprof }}
```
Comment (manno) [inline charts/fleet/templates/deployment.yaml:62]: I think this is obsolete now, /tmp is already a volume. As far as I can tell `/tmp/pprof` is fine. @moio ?

File: pkg/controllers/image/image.go [hunk 2]
```diff
@@ -278,6 +279,32 @@ func (h handler) onChangeGitRepo(gitrepo *v1alpha1.GitRepo, status v1alpha1.GitR
 	return status, err
 }
 
+func (h handler) setupKnownHosts(gitrepo *v1alpha1.GitRepo, data []byte) error {
+	tmpdir, err := os.MkdirTemp("", gitrepo.Namespace+"-"+gitrepo.Name)
```
Comment (manno) [inline pkg/controllers/image/image.go:283]: I wouldn't use user input here to construct the temp dir, should be safe and a [random suffix](https://pkg.go.dev/os#MkdirTemp) is added, but still.

```suggestion
	tmpdir, err := os.MkdirTemp("", "fleet-")
```

Comment (manno) [inline pkg/controllers/image/image.go:283]: We should probably remove that dir when the operation exits?

**Review by manno — CHANGES_REQUESTED** [review]

File: charts/fleet/templates/deployment.yaml (hunk 1 above)
Comment (manno) [reply charts/fleet/templates/deployment.yaml:62]: Ah thanks. I didn't expand the diff :) 👍

File: pkg/controllers/image/image.go (hunk 2 above)
Comment (olblak) [reply pkg/controllers/image/image.go:283]: Is there any chance to have some sort of collusion here? Like two different process fetching the digest for the same tag at a slightly different type which could lead to different digest for the same tag?

File: pkg/controllers/image/image.go [hunk 3]
```diff
@@ -278,6 +287,32 @@ func (h handler) onChangeGitRepo(gitrepo *v1alpha1.GitRepo, status v1alpha1.GitR
...
 }
 
+func (h handler) setupKnownHosts(gitrepo *v1alpha1.GitRepo, data []byte) error {
+	tmpdir, err := os.MkdirTemp("", fmt.Sprintf("ssh-%s-%s-", gitrepo.Namespace, gitrepo.Name))
+	if err != nil {
+		return err
+	}
+
+	known := path.Join(tmpdir, "known_hosts")
+	err = os.Setenv("SSH_KNOWN_HOSTS", known)
```
Comment (manno) [inline pkg/controllers/image/image.go:297]: The controller is started from a goroutine (I think this code in lasso starts controllers: https://github.com/rancher/lasso/blob/master/pkg/controller/controller.go#L121-L156).

Goroutines share environment variables, as they are the in the same OS process.  I couldn't find any explicit documentation for this, however https://go.dev/play/p/e0L5ixUe4qL illustrates the issue.

By setting the variable `SSH_KNOWN_HOSTS` to a temp file with a random name, we could in theory fail a concurrent second image scan. I think this isn't too unlikely if the interval is low enough and the SSH connection is slow enough.

As a workaround, we could add a lock around this code (which is not good for performance) or pass the known hosts to go-git without the env var somehow?

File: pkg/controllers/image/image.go [hunk 4]
```diff
@@ -278,6 +287,32 @@ func (h handler) onChangeGitRepo(gitrepo *v1alpha1.GitRepo, status v1alpha1.GitR
 	return status, err
 }
 
+func (h handler) setupKnownHosts(gitrepo *v1alpha1.GitRepo, data []byte) error {
+	tmpdir, err := os.MkdirTemp("", fmt.Sprintf("ssh-%s-%s-", gitrepo.Namespace, gitrepo.Name))
```
Comment (manno) [inline pkg/controllers/image/image.go:291]:
```suggestion
	tmpdir, err := os.MkdirTemp("", fmt.Sprintf("fleet-ssh-%s-%s-", gitrepo.Namespace, gitrepo.Name))
```

So we can easily remove the dirs locally, without interfering with other temp files?

**Review by manno — APPROVED** [review]

**Pull request merged**